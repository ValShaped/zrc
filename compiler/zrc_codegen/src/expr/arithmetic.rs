//! code generation for arithmetic expressions

use inkwell::{
    builder::BuilderError,
    values::{BasicValue, BasicValueEnum, IntValue},
};
use zrc_typeck::tast::{
    expr::{Arithmetic, BinaryBitwise, TypedExpr},
    ty::Type,
};

use crate::{
    bb::{BasicBlockAnd, BasicBlockExt},
    ctx::BlockCtx,
    expr::{CgExprArgs, cg_expr},
    ty::llvm_basic_type,
    unpack,
};

/// Build the required instruction for a [`BinaryBitwise`] operation
pub fn build_binary_bitwise<'ctx>(
    cg: BlockCtx<'ctx, '_, '_>,
    op: BinaryBitwise,
    lhs: IntValue<'ctx>,
    rhs: IntValue<'ctx>,
    result_is_signed: bool,
) -> Result<IntValue<'ctx>, BuilderError> {
    match op {
        BinaryBitwise::And => cg.builder.build_and(lhs, rhs, "and"),
        BinaryBitwise::Or => cg.builder.build_or(lhs, rhs, "or"),
        BinaryBitwise::Xor => cg.builder.build_xor(lhs, rhs, "xor"),
        BinaryBitwise::Shl => cg.builder.build_left_shift(lhs, rhs, "shl"),
        BinaryBitwise::Shr => cg
            .builder
            .build_right_shift(lhs, rhs, result_is_signed, "shr"),
    }
}
/// Build the required instruction for an [`Arithmetic`] operation
pub fn build_arithmetic<'ctx>(
    cg: BlockCtx<'ctx, '_, '_>,
    op: Arithmetic,
    lhs: IntValue<'ctx>,
    rhs: IntValue<'ctx>,
    result_is_signed: bool,
) -> Result<IntValue<'ctx>, BuilderError> {
    match op {
        Arithmetic::Addition => cg.builder.build_int_add(lhs, rhs, "add"),
        Arithmetic::Subtraction => cg.builder.build_int_sub(lhs, rhs, "sub"),
        Arithmetic::Multiplication => cg.builder.build_int_mul(lhs, rhs, "mul"),

        Arithmetic::Division if result_is_signed => {
            cg.builder.build_int_signed_div(lhs, rhs, "div")
        }
        Arithmetic::Division => cg.builder.build_int_unsigned_div(lhs, rhs, "div"),

        Arithmetic::Modulo if result_is_signed => cg.builder.build_int_signed_rem(lhs, rhs, "rem"),
        Arithmetic::Modulo => cg.builder.build_int_unsigned_rem(lhs, rhs, "rem"),
    }
}

/// Code generate a binary bitwise operation
pub fn cg_binary_bitwise<'ctx, 'input>(
    CgExprArgs {
        cg,
        mut bb,
        inferred_type,
        ..
    }: CgExprArgs<'ctx, 'input, '_>,
    op: BinaryBitwise,
    lhs: Box<TypedExpr<'input>>,
    rhs: Box<TypedExpr<'input>>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let lhs = unpack!(bb = cg_expr(cg, bb, *lhs));
    let rhs = unpack!(bb = cg_expr(cg, bb, *rhs));

    let reg = build_binary_bitwise(
        cg,
        op,
        lhs.into_int_value(),
        rhs.into_int_value(),
        inferred_type.is_signed_integer(),
    )
    .expect("binary bitwise operation should have compiled successfully");

    bb.and(reg.as_basic_value_enum())
}

/// Code generate an arithmetic operation
pub fn cg_arithmetic<'ctx, 'input>(
    CgExprArgs {
        cg,
        mut bb,
        inferred_type,
        ..
    }: CgExprArgs<'ctx, 'input, '_>,
    op: Arithmetic,
    lhs: Box<TypedExpr<'input>>,
    rhs: Box<TypedExpr<'input>>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let lhs_ty = lhs.inferred_type.clone();
    let lhs = unpack!(bb = cg_expr(cg, bb, *lhs));
    let rhs = unpack!(bb = cg_expr(cg, bb, *rhs));

    if let Type::Ptr(pointee) = lhs_ty {
        // Most languages make incrementing a pointer increase the address by the size
        // of the pointee type, hence our use of `gep`.
        // REVIEW: Is this the approach we want to take?
        #[allow(clippy::wildcard_enum_match_arm)]
        let reg = match op {
            // SAFETY: This can segfault if indices are used incorrectly
            // This is only used for pointer arithmetic, so the indices should be correct
            Arithmetic::Addition => unsafe {
                cg.builder.build_gep(
                    llvm_basic_type(&cg, &pointee).0,
                    lhs.into_pointer_value(),
                    &[rhs.into_int_value()],
                    "ptr_add",
                )
            },
            Arithmetic::Subtraction => {
                let rhs = cg
                    .builder
                    .build_int_neg(rhs.into_int_value(), "neg")
                    .expect("negation should have compiled successfully")
                    .as_basic_value_enum();

                // SAFETY: This can segfault if indices are used incorrectly
                // This is only used for pointer arithmetic, so the indices should be
                // correct
                unsafe {
                    cg.builder.build_gep(
                        llvm_basic_type(&cg, &pointee).0,
                        lhs.into_pointer_value(),
                        &[rhs.into_int_value()],
                        "ptr_sub",
                    )
                }
            }
            _ => panic!("invalid pointer arithmetic operation"),
        }
        .expect("pointer arithmetic should have compiled successfully");

        bb.and(reg.as_basic_value_enum())
    } else {
        let reg = build_arithmetic(
            cg,
            op,
            lhs.into_int_value(),
            rhs.into_int_value(),
            inferred_type.is_signed_integer(),
        )
        .expect("arithmetic operation should have compiled successfully");

        bb.and(reg.as_basic_value_enum())
    }
}

/// Code generate a unary bitwise NOT operation
pub fn cg_unary_bitwise_not<'ctx, 'input>(
    CgExprArgs { cg, mut bb, .. }: CgExprArgs<'ctx, 'input, '_>,
    x: Box<TypedExpr<'input>>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let value = unpack!(bb = cg_expr(cg, bb, *x));

    let reg = cg
        .builder
        .build_not(value.into_int_value(), "not")
        .expect("not should have compiled successfully");

    bb.and(reg.as_basic_value_enum())
}

/// Code generate a unary minus operation
pub fn cg_unary_minus<'ctx, 'input>(
    CgExprArgs { cg, mut bb, .. }: CgExprArgs<'ctx, 'input, '_>,
    x: Box<TypedExpr<'input>>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let value = unpack!(bb = cg_expr(cg, bb, *x));

    let reg = cg
        .builder
        .build_int_neg(value.into_int_value(), "neg")
        .expect("negation should have compiled successfully");

    bb.and(reg.as_basic_value_enum())
}

#[cfg(test)]
mod tests {
    // Please read the "Common patterns in tests" section of crate::test_utils for
    // more information on how code generator tests are structured.

    use indoc::indoc;

    use crate::cg_snapshot_test;

    #[test]
    fn pointer_arithmetic_generates_proper_gep() {
        cg_snapshot_test!(indoc! {r#"
                fn test() {
                    let x: *i32;

                    // TEST: should create a GEP that is the same as "x[4]"
                    let y = x + 4 as usize;
                    // TEST: and the same, with -4:
                    let z = x - 4 as usize;
                }
            "#});
    }

    #[test]
    fn arithmetic_operators_generate() {
        cg_snapshot_test!(indoc! {"
                fn get_int() -> i32;
                fn get_uint() -> u32;

                fn test() {
                    let sx = get_int();
                    let sy = get_int();
                    let ux = get_uint();
                    let uy = get_uint();

                    // TEST: should create an `add i32` instruction
                    let s_add = sx + sy;
                    let u_add = ux + uy;

                    // TEST: should create a `sub i32` instruction
                    let s_sub = sx - sy;
                    let u_sub = ux - uy;

                    // TEST: should create a `mul i32` instruction
                    let s_mul = sx * sy;
                    let u_mul = ux * uy;

                    // TEST: should create a `sdiv i32` instruction
                    let s_div = sx / sy;
                    // TEST: should create a `udiv i32` instruction
                    let u_div = ux / uy;

                    // TEST: should create a `srem i32` instruction
                    let s_rem = sx % sy;
                    // TEST: should create a `urem i32` instruction
                    let u_rem = ux % uy;
                }
            "});
    }

    #[test]
    fn bitwise_operators_generate() {
        cg_snapshot_test!(indoc! {"
                fn get_int() -> i32;
                fn get_uint() -> u32;

                fn test() {
                    let x = get_int();
                    let y = get_int();
                    let u = get_uint();

                    // TEST: should create a `not i32` instruction
                    let not = ~x;

                    // TEST: should create a `and i32` instruction
                    let and = x & y;

                    // TEST: should create a `or i32` instruction
                    let or = x | y;

                    // TEST: should create a `xor i32` instruction
                    let xor = x ^ y;

                    // TEST: should create a `shl i32` instruction
                    let shl = x << u;

                    // TEST: should create a `lshr i32` instruction
                    let lshr = u >> u;

                    // TEST: should create a `ashr i32` instruction (as the lhs is signed)
                    let ashr = x >> u;
                }
            "});
    }
}
