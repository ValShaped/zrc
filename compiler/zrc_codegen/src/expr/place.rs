//! Code generation for Place types

use inkwell::{
    basic_block::BasicBlock,
    debug_info::AsDIScope,
    values::{BasicValue, PointerValue},
};
use zrc_typeck::tast::{
    expr::{Place, PlaceKind},
    ty::Type,
};

use super::cg_expr;
use crate::{
    bb::{BasicBlockAnd, BasicBlockExt},
    ctx::BlockCtx,
    ty::llvm_basic_type,
    unpack,
};

/// Resolve a place to its pointer
#[allow(clippy::too_many_arguments, clippy::too_many_lines)]
pub fn cg_place<'ctx>(
    cg: BlockCtx<'ctx, '_, '_>,
    mut bb: BasicBlock<'ctx>,
    place: Place,
) -> BasicBlockAnd<'ctx, PointerValue<'ctx>> {
    let place_span = place.kind.span();
    let line_and_col = cg.line_lookup.lookup_from_index(place_span.start());
    let debug_location = cg.dbg_builder.create_debug_location(
        cg.ctx,
        line_and_col.line,
        line_and_col.col,
        cg.dbg_scope.as_debug_info_scope(),
        None,
    );
    cg.builder.set_current_debug_location(debug_location);

    match place.kind.into_value() {
        PlaceKind::Variable(x) => {
            let reg = cg
                .scope
                .get(x)
                .expect("identifier that passed typeck should exist in the CgScope");

            bb.and(reg)
        }

        PlaceKind::Deref(x) => {
            let value = unpack!(bb = cg_expr(cg, bb, *x));

            bb.and(value.into_pointer_value())
        }

        PlaceKind::Index(ptr, idx) => {
            let ptr = unpack!(bb = cg_expr(cg, bb, *ptr));
            let idx = unpack!(bb = cg_expr(cg, bb, *idx));

            // SAFETY: If indices are used incorrectly this may segfault
            // TODO: Is this actually safely used?
            let reg = unsafe {
                cg.builder.build_gep(
                    llvm_basic_type(&cg, &place.inferred_type).0,
                    ptr.into_pointer_value(),
                    &[idx.into_int_value()],
                    "gep",
                )
            }
            .expect("building GEP instruction should succeed");

            bb.and(reg.as_basic_value_enum().into_pointer_value())
        }

        #[allow(clippy::wildcard_enum_match_arm)]
        PlaceKind::Dot(x, prop) => match &x.inferred_type {
            Type::Struct(contents) => {
                let x_ty = llvm_basic_type(&cg, &x.inferred_type).0;
                let prop_idx = contents
                    .iter()
                    .position(|(got_key, _)| *got_key == prop.into_value())
                    .expect("invalid struct field");

                let x = unpack!(bb = cg_place(cg, bb, *x));

                let reg = cg
                    .builder
                    .build_struct_gep(
                        x_ty,
                        x,
                        prop_idx
                            .try_into()
                            .expect("got more than u32::MAX as key index? HOW?"),
                        "gep",
                    )
                    .expect("building GEP instruction should succeed");

                bb.and(reg.as_basic_value_enum().into_pointer_value())
            }
            Type::Union(_) => {
                // All we need to do is cast the pointer, but there's no `bitcast` anymore,
                // so just return it and it'll take on the correct type

                let value = unpack!(bb = cg_place(cg, bb, *x));

                bb.and(value)
            }
            _ => panic!("cannot access property of non-struct"),
        },
    }
}

#[cfg(test)]
mod tests {
    // Please read the "Common patterns in tests" section of crate::test_utils for
    // more information on how code generator tests are structured.

    use indoc::indoc;

    use crate::cg_snapshot_test;

    // Remember: In all of these tests, cg_place returns a *pointer* to the data in
    // the place.

    #[test]
    fn basic_identifiers_in_place_position() {
        cg_snapshot_test!(indoc! {"
                fn test() {
                    let x = 6;

                    // TEST: we should simply be `store`ing to the %let_x we created
                    x = 7;
                }
            "});
    }

    #[test]
    fn identifier_deref_generates_as_expected() {
        cg_snapshot_test!(indoc! {"
                fn test() {
                    let x: *i32;

                    // TEST: x is *i32, so %let_x is **i32 (ptr to the stack). we should load from
                    // %let_x to obtain the actual pointer. we should then store to that result
                    // value (we should never load it)
                    *x = 4;
                }
            "});
    }

    #[test]
    fn other_deref_generates_as_expected() {
        cg_snapshot_test!(indoc! {"
                fn test() {
                    // TEST: because cg_place returns a *pointer* to the represented value, handling
                    // *5 in a place context should return the address of *5, which is &*5 = 5.
                    // for this reason, we should literally be `store`ing to the hardcoded address
                    // 5, and never *loading* from it (because if we do load we may not be actually
                    // writing to that address)
                    // we use 5 not 0 because 0 is just 'ptr null'
                    *(5 as *i32) = 0;
                }
            "});
    }

    #[test]
    fn pointer_indexing_in_place_position() {
        cg_snapshot_test!(indoc! {"
                fn test() {
                    let x: *i32;

                    // TEST: `x` is *i32, so %let_x is a **i32 (ptr to the stack).
                    // %let_x needs to be GEP'd into and then stored into, but we must not load
                    // from the address.
                    x[4 as usize] = 5;
                }
            "});
    }

    #[test]
    fn struct_property_access_in_place_position() {
        cg_snapshot_test!(indoc! {"
                struct S { x: i32, y: i32 }

                fn test() {
                    let x: S;

                    // TEST: the value must NOT be loaded! it must simply gep to obtain a pointer,
                    // then `store` into that pointer.
                    x.y = 4;
                }
            "});
    }

    #[test]
    fn union_property_access_in_place_position() {
        cg_snapshot_test!(indoc! {"
                union U { x: i32, y: i8 }

                fn test() {
                    let x: U;

                    // TEST: the pointer is cast and then written to as an i32
                    x.x = 4;

                    // TEST: the pointer is cast and then written to as an i8
                    x.y = 5 as i8;
                }
            "});
    }
}
