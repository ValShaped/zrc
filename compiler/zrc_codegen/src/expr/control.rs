//! code generation for control flow expressions

use inkwell::values::{BasicValue, BasicValueEnum};
use zrc_typeck::tast::expr::{Place, TypedExpr};

use super::place::cg_place;
use crate::{
    bb::{BasicBlockAnd, BasicBlockExt},
    expr::{CgExprArgs, cg_expr},
    ty::{llvm_basic_type, llvm_type},
    unpack,
};

/// Code generate a comma expression
pub fn cg_comma<'ctx, 'input>(
    CgExprArgs { cg, mut bb, .. }: CgExprArgs<'ctx, 'input, '_>,
    lhs: Box<TypedExpr<'input>>,
    rhs: Box<TypedExpr<'input>>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let _ = unpack!(bb = cg_expr(cg, bb, *lhs));
    cg_expr(cg, bb, *rhs)
}

/// Code generate a function call expression
pub fn cg_call<'ctx, 'input>(
    CgExprArgs { cg, mut bb, .. }: CgExprArgs<'ctx, 'input, '_>,
    f: Place<'input>,
    args: Vec<TypedExpr<'input>>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let llvm_f_type = llvm_type(&cg, &f.inferred_type).0.into_function_type();

    // will always be a function pointer
    let f_ptr = unpack!(bb = cg_place(cg, bb, f));

    let mut bb = bb;
    let old_args = args;
    let mut args = vec![];
    for arg in old_args {
        let new_arg = unpack!(bb = cg_expr(cg, bb, arg));
        args.push(new_arg.into());
    }

    let ret = cg
        .builder
        .build_indirect_call(llvm_f_type, f_ptr, &args, "call")
        .expect("call should have compiled successfully");

    bb.and(if ret.try_as_basic_value().is_left() {
        ret.try_as_basic_value().unwrap_left()
    } else {
        cg.ctx.i8_type().get_undef().as_basic_value_enum()
    })
}

/// Code generate a ternary expression
pub fn cg_ternary<'ctx, 'input>(
    CgExprArgs {
        cg,
        bb,
        inferred_type,
        ..
    }: CgExprArgs<'ctx, 'input, '_>,
    cond: Box<TypedExpr<'input>>,
    lhs: Box<TypedExpr<'input>>,
    rhs: Box<TypedExpr<'input>>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let cond = cg_expr(cg, bb, *cond).into_value();

    // If lhs and rhs are registers, the code generated will look like:
    //   entry:
    //       ...
    //       %cond = ...
    //       br i1 %cond, label %if_true, label %if_false
    //   if_true:
    //       ... yields %lhs
    //       br label %end
    //   if_false:
    //       ... yields %rhs
    //       br label %end
    //   end:
    //       %yield = phi TY [ %lhs, %if_true ], [ %rhs, %if_false ]
    //
    // If they are constants, the code generated will load as:
    //   entry:
    //       ...
    //       %cond = ...
    //       br i1 %cond, label %if_true, label %if_false
    //   if_true:
    //       br label %end
    //   if_false:
    //       br label %end
    //   end:
    //       %yield = phi TY [ LHS VALUE, %if_true ], [ RHS VALUE, %if_false ]
    //
    // This is just a weird way to `select` and will be inlined.

    let mut if_true_bb = cg.ctx.append_basic_block(cg.fn_value, "if_true");
    let mut if_false_bb = cg.ctx.append_basic_block(cg.fn_value, "if_false");
    let end_bb = cg.ctx.append_basic_block(cg.fn_value, "end");

    cg.builder
        .build_conditional_branch(cond.into_int_value(), if_true_bb, if_false_bb)
        .expect("conditional branch should have been created successfully");

    cg.builder.position_at_end(if_true_bb);
    let if_true = unpack!(if_true_bb = cg_expr(cg, if_true_bb, *lhs));
    cg.builder
        .build_unconditional_branch(end_bb)
        .expect("unconditional branch should have been created successfully");

    cg.builder.position_at_end(if_false_bb);
    let if_false = unpack!(if_false_bb = cg_expr(cg, if_false_bb, *rhs));
    cg.builder
        .build_unconditional_branch(end_bb)
        .expect("unconditional branch should have been created successfully");

    cg.builder.position_at_end(end_bb);
    let result_reg = cg
        .builder
        .build_phi(llvm_basic_type(&cg, &inferred_type).0, "yield")
        .expect("phi node should have been created successfully");
    result_reg.add_incoming(&[(&if_true, if_true_bb), (&if_false, if_false_bb)]);

    end_bb.and(result_reg.as_basic_value())
}

#[cfg(test)]
mod tests {
    // Please read the "Common patterns in tests" section of crate::test_utils for
    // more information on how code generator tests are structured.

    use indoc::indoc;

    use crate::cg_snapshot_test;

    #[test]
    fn comma_yields_right_value() {
        cg_snapshot_test!(indoc! {"
                fn f();
                fn g() -> i32;

                fn test() -> i32 {
                    // TEST: f() is run, g() is run and used as the return value
                    return f(), g();
                }
            "});
    }

    #[test]
    fn ternary_operations_generate() {
        cg_snapshot_test!(indoc! {"
                fn get_bool() -> bool;
                fn get_int() -> i32;
                fn take_int(x: i32);
                fn test() {
                    // TEST: should produce a proper diamond-shaped cfg
                    let num = get_bool() ? get_int() : 3;
                    take_int(num);
                }
            "});
    }
}
