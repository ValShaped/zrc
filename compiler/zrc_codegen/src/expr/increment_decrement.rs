//! code generation for prefix/postfix increment/decrement expressions

use inkwell::values::{BasicValue, BasicValueEnum};
use zrc_typeck::tast::expr::Place;

use super::place::cg_place;
use crate::{
    bb::{BasicBlockAnd, BasicBlockExt},
    expr::CgExprArgs,
    ty::{llvm_basic_type, llvm_int_type},
    unpack,
};

/// Code generate a prefix increment expression
pub fn cg_prefix_increment<'ctx, 'input>(
    CgExprArgs {
        cg,
        mut bb,
        inferred_type,
        ..
    }: CgExprArgs<'ctx, 'input, '_>,
    place: Place<'input>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let place_ptr = unpack!(bb = cg_place(cg, bb, place));

    // Load current value
    let current = cg
        .builder
        .build_load(llvm_basic_type(&cg, &inferred_type).0, place_ptr, "load")
        .expect("prefix increment load should have compiled successfully");

    // Add 1
    let one = llvm_int_type(&cg, &inferred_type).0.const_int(1, false);
    let new_value = cg
        .builder
        .build_int_add(current.into_int_value(), one, "inc")
        .expect("prefix increment add should have compiled successfully");

    // Store back
    cg.builder
        .build_store(place_ptr, new_value)
        .expect("prefix increment store should have compiled successfully");

    // Return new value
    bb.and(new_value.as_basic_value_enum())
}

/// Code generate a prefix decrement expression
pub fn cg_prefix_decrement<'ctx, 'input>(
    CgExprArgs {
        cg,
        mut bb,
        inferred_type,
        ..
    }: CgExprArgs<'ctx, 'input, '_>,
    place: Place<'input>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let place_ptr = unpack!(bb = cg_place(cg, bb, place));

    // Load current value
    let current = cg
        .builder
        .build_load(llvm_basic_type(&cg, &inferred_type).0, place_ptr, "load")
        .expect("prefix decrement load should have compiled successfully");

    // Subtract 1
    let one = llvm_int_type(&cg, &inferred_type).0.const_int(1, false);
    let new_value = cg
        .builder
        .build_int_sub(current.into_int_value(), one, "dec")
        .expect("prefix decrement sub should have compiled successfully");

    // Store back
    cg.builder
        .build_store(place_ptr, new_value)
        .expect("prefix decrement store should have compiled successfully");

    // Return new value
    bb.and(new_value.as_basic_value_enum())
}

/// Code generate a postfix increment expression
pub fn cg_postfix_increment<'ctx, 'input>(
    CgExprArgs {
        cg,
        mut bb,
        inferred_type,
        ..
    }: CgExprArgs<'ctx, 'input, '_>,
    place: Place<'input>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let place_ptr = unpack!(bb = cg_place(cg, bb, place));

    // Load current value
    let current = cg
        .builder
        .build_load(llvm_basic_type(&cg, &inferred_type).0, place_ptr, "load")
        .expect("postfix increment load should have compiled successfully");

    // Add 1
    let one = llvm_int_type(&cg, &inferred_type).0.const_int(1, false);
    let new_value = cg
        .builder
        .build_int_add(current.into_int_value(), one, "inc")
        .expect("postfix increment add should have compiled successfully");

    // Store back
    cg.builder
        .build_store(place_ptr, new_value)
        .expect("postfix increment store should have compiled successfully");

    // Return old value
    bb.and(current)
}

/// Code generate a postfix decrement expression
pub fn cg_postfix_decrement<'ctx, 'input>(
    CgExprArgs {
        cg,
        mut bb,
        inferred_type,
        ..
    }: CgExprArgs<'ctx, 'input, '_>,
    place: Place<'input>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let place_ptr = unpack!(bb = cg_place(cg, bb, place));

    // Load current value
    let current = cg
        .builder
        .build_load(llvm_basic_type(&cg, &inferred_type).0, place_ptr, "load")
        .expect("postfix decrement load should have compiled successfully");

    // Subtract 1
    let one = llvm_int_type(&cg, &inferred_type).0.const_int(1, false);
    let new_value = cg
        .builder
        .build_int_sub(current.into_int_value(), one, "dec")
        .expect("postfix decrement sub should have compiled successfully");

    // Store back
    cg.builder
        .build_store(place_ptr, new_value)
        .expect("postfix decrement store should have compiled successfully");

    // Return old value
    bb.and(current)
}
