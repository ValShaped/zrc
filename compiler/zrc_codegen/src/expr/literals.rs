//! code generation for literal expressions

use inkwell::{
    types::StringRadix,
    values::{BasicValue, BasicValueEnum},
};
use zrc_typeck::tast::expr::{NumberLiteral, Place, PlaceKind, StringTok, ZrcString};
use zrc_utils::span::Spannable;

use super::place::cg_place;
use crate::{
    bb::{BasicBlockAnd, BasicBlockExt},
    expr::CgExprArgs,
    ty::{llvm_basic_type, llvm_int_type},
    unpack,
};

/// Get a [`NumberLiteral`]'s [`StringRadix`]
const fn number_literal_radix(n: &NumberLiteral) -> StringRadix {
    match n {
        NumberLiteral::Decimal(_) => StringRadix::Decimal,
        NumberLiteral::Binary(_) => StringRadix::Binary,
        NumberLiteral::Hexadecimal(_) => StringRadix::Hexadecimal,
    }
}

/// Generate LLVM IR for a number literal
pub fn cg_number_literal<'ctx, 'input>(
    CgExprArgs {
        cg,
        bb,
        inferred_type,
        ..
    }: CgExprArgs<'ctx, 'input, '_>,
    n: &NumberLiteral<'input>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let no_underscores = n.text_content().replace('_', "");

    bb.and(
        llvm_int_type(&cg, &inferred_type)
            .0
            .const_int_from_string(&no_underscores, number_literal_radix(n))
            .expect("number literal should have parsed correctly")
            .as_basic_value_enum(),
    )
}

/// Generate LLVM IR for a string literal
pub fn cg_string_literal<'ctx, 'input>(
    CgExprArgs { cg, bb, .. }: CgExprArgs<'ctx, 'input, '_>,
    str: &ZrcString<'input>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    bb.and(
        cg.builder
            .build_global_string_ptr(&str.as_bytes(), "str")
            .expect("string should have built successfully")
            .as_basic_value_enum(),
    )
}

/// Generate LLVM IR for a char literal
pub fn cg_char_literal<'ctx, 'input>(
    CgExprArgs { cg, bb, .. }: CgExprArgs<'ctx, 'input, '_>,
    ch: &StringTok<'input>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    bb.and(
        cg.ctx
            .i8_type()
            .const_int(ch.as_byte().into(), false)
            .as_basic_value_enum(),
    )
}

/// Generate LLVM IR for a boolean literal
pub fn cg_boolean_literal<'ctx>(
    CgExprArgs { cg, bb, .. }: CgExprArgs<'ctx, '_, '_>,
    value: bool,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    bb.and(
        cg.ctx
            .bool_type()
            .const_int(value.into(), false)
            .as_basic_value_enum(),
    )
}

/// Generate LLVM IR for an identifier
pub fn cg_identifier<'ctx, 'input>(
    CgExprArgs {
        cg,
        mut bb,
        expr_span,
        inferred_type,
    }: CgExprArgs<'ctx, 'input, '_>,
    id: &'input str,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let place = unpack!(
        bb = cg_place(
            cg,
            bb,
            Place {
                inferred_type: inferred_type.clone(),
                kind: PlaceKind::Variable(id).in_span(expr_span),
            },
        )
    );

    let reg = cg
        .builder
        .build_load(llvm_basic_type(&cg, &inferred_type).0, place, "load")
        .expect("ident load should have built successfully");

    bb.and(reg.as_basic_value_enum())
}

#[cfg(test)]
mod tests {
    // Please read the "Common patterns in tests" section of crate::test_utils for
    // more information on how code generator tests are structured.

    use indoc::indoc;

    use crate::cg_snapshot_test;

    #[test]
    fn typed_integers_generate_properly() {
        cg_snapshot_test!(indoc! {"
                fn test() -> i8 {
                    // TEST: returns `i8`
                    return 4i8;
                }
            "});
    }

    #[test]
    fn string_literal_escapes_generate() {
        cg_snapshot_test!(indoc! {r#"
                fn test() {
                    // TEST: should properly generate \xNN for each escape
                    let x = "\n\r\t\\\"\x41\u{2400}\0";
                }
            "#});
    }

    /// Tests to ensure non-decimal integer literals
    /// 1. don't panic
    /// 2. are valid.
    ///
    /// <https://github.com/zirco-lang/zrc/issues/119>
    #[test]
    fn regression_119_special_integer_literals() {
        cg_snapshot_test!(indoc! {"
                fn test() {
                    let a = 0b10_10;
                    let b = 0x1F_A4;
                }
            "});
    }
}
