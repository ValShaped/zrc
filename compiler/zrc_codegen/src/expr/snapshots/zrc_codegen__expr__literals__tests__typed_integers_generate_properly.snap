---
source: compiler/zrc_codegen/src/expr/literals.rs
description: "fn test() -> i8 {\n    // TEST: returns `i8`\n    return 4i8;\n}\n"
expression: resulting_ir
---
; ModuleID = 'test.zr'
source_filename = "test.zr"

define i8 @test() !dbg !3 {
entry:
  ret i8 4, !dbg !7
}

!llvm.module.flags = !{!0}
!llvm.dbg.cu = !{!1}

!0 = !{i32 2, !"Debug Info Version", i32 3}
!1 = distinct !DICompileUnit(language: DW_LANG_C, file: !2, producer: "zrc test runner", isOptimized: false, flags: "zrc --fake-args", runtimeVersion: 0, emissionKind: FullDebug, splitDebugInlining: false)
!2 = !DIFile(filename: "test.zr", directory: "/fake/path")
!3 = distinct !DISubprogram(name: "test", linkageName: "test", scope: null, file: !2, line: 1, type: !4, scopeLine: 1, spFlags: DISPFlagDefinition, unit: !1)
!4 = !DISubroutineType(types: !5)
!5 = !{!6}
!6 = !DIBasicType(name: "i8")
!7 = !DILocation(line: 3, column: 12, scope: !8)
!8 = distinct !DILexicalBlock(scope: !9, file: !2, line: 1, column: 17)
!9 = distinct !DILexicalBlock(scope: !3, file: !2, line: 1, column: 17)
